src/config_generator/reader.rs

```diff
@@ -0,0 +1,111 @@
+use anyhow::Result;
+
+use crate::config::{Config, Link, LinkType};
+use crate::config_generator::from_proto::from_proto;
+use crate::config_generator::source::GeneratorSource;
+use crate::merge_right::MergeRight;
+use crate::proto_reader::ProtoReader;
+use crate::runtime::TargetRuntime;
+
+pub struct GeneratorReader {
+    proto_reader: ProtoReader,
+}
+impl GeneratorReader {
+    pub fn init(runtime: TargetRuntime) -> Self {
+        Self { proto_reader: ProtoReader::init(runtime) }
+    }
+
+    pub async fn read_all<T: AsRef<str>>(&self, files: &[T], query: &str) -> Result<Config> {
+        let mut links = vec![];
+        let proto_metadata = self.proto_reader.read_all(files).await?;
+
+        let mut config = Config::default();
+
+        for metadata in proto_metadata {
+            let source = match metadata.file_read.content_ty {
+                Some(content_ty) => match GeneratorSource::detect(content_ty.as_str()) {
+                    Ok(source) => source,
+                    Err(_) => GeneratorSource::detect(metadata.file_read.path.as_str())?,
+                },
+                None => GeneratorSource::detect(metadata.file_read.path.as_str())?,
+            };
+            match source {
+                GeneratorSource::PROTO => {
+                    links.push(Link {
+                        id: None,
+                        src: metadata.file_read.path.clone(),
+                        type_of: LinkType::Protobuf,
+                    });
+                    config = config.merge_right(from_proto(&[metadata.descriptor_set], query));
+                }
+            }
+        }
+
+        config.links = links;
+        Ok(config)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::path::PathBuf;
+
+    use super::*;
+
+    fn start_mock_server() -> httpmock::MockServer {
+        httpmock::MockServer::start()
+    }
+
+    #[tokio::test]
+    async fn test_read_all() {
+        let server = start_mock_server();
+        let runtime = crate::runtime::test::init(None);
+        let test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("src")
+            .join("config_generator")
+            .join("proto");
+
```

Comment:
instead of calling `join` multiple times, just pass `src/config_generator/proto`.